### PR TITLE
Persistent connections pool

### PR DESCRIPTION
**Pull request by yatsukhnenko**
The work is not finished yet but the main idea is clear :)

### CODE REVIEW

File: redis.c [hunk 1]
```diff
@@ -53,6 +53,8 @@ zend_class_entry *redis_exception_ce;
 extern zend_function_entry redis_array_functions[];
 extern zend_function_entry redis_cluster_functions[];
 
+int le_pconnect;
```
Comment (michael-grunder) [inline redis.c]: Maybe a more unique name just to avoid potential conflict?  Maybe `le_redis_pconnect`?

File: library.c [hunk 2]
```diff
@@ -1869,6 +1883,36 @@ redis_sock_disconnect(RedisSock *redis_sock, int force TSRMLS_DC)
...
+                } else {
+                    const char *fmtstr = "%s:%d";
+#ifdef HAVE_IPV6
+                    if (strchr(ZSTR_VAL(redis_sock->host), ':') != NULL) {
+                        fmtstr = "[%s]:%d";
+                    }
+#endif
+                    snprintf(host, sizeof(host), fmtstr, ZSTR_VAL(redis_sock->host), redis_sock->port);
+                }
+                if (redis_sock->persistent_id) {
```
Comment (michael-grunder) [inline library.c]: I could be wrong, but I think with proper connection pooling there is no longer any need for directing the `persistent_id`, or by creating it with the timeout.  If we're separating connections in a pool we probably just want to hash by `host:port` or the fully qualified unix socket.

File: redis.c (hunk 1 above)
Comment (yatsukhnenko) [reply redis.c]: Good idea! I'll change it.

File: library.c (hunk 2 above)
Comment (yatsukhnenko) [reply library.c]: Seems reasonable. I changed persistend_id for pool resource and remove it for sockets

**Comment by yatsukhnenko** [comment]
> 1. Do we need to introduce the feature disabled by default.

There is nothing "breaking" here but to protect us from swearing users I thinks we need to make this feature disabled by default in version supported PHP5 and PHP7 but enabled in version for PHP7 only.
In any way users need to be able to turn it on/off via connection option or INI setting.

**Comment by yatsukhnenko** [comment]
> 2\. Do we need to have configurable limits like the MySQL extension has

I'm not sure about this limit. If OS allows to create one more connection why should we deny? But I'm not against this :smile:

**Comment by michael-grunder** [comment]
I'm pretty sure people will want to be able to set limits on the total number of persistent connections, but we could allow `-1` to mean "unlimited".

Also, I don't know if we need to use the [timeout](https://github.com/phpredis/phpredis/pull/1506/files#diff-c5867178362a2abccbf1419f7363735cR1808) anymore in the persistent ID.  That was created before I worked on phpredis but it shouldn't matter when operating against a connection pool because that's a "connection timeout", and theoretically we're already connected.

**Comment by yatsukhnenko** [comment]
I'm thinking about counter which will be incremented for every new connection

**Comment by yatsukhnenko** [comment]
And for limits I think we need settings for min and max number of connections in pool (like php-fpm pool has min and max number of child proccesses)

**Comment by michael-grunder** [comment]
> And for limits I think we need settings for min and max number of connections in pool (like php-fpm pool has min and max number of child processes)

I think we just need a maximum number of connections and we can use a global variable for that.

**Comment by yatsukhnenko** [comment]
> Also, I don't know if we need to use the [timeout](https://github.com/phpredis/phpredis/pull/1506/files#diff-c5867178362a2abccbf1419f7363735cR1808) anymore in the persistent ID.

At this place timeout used for creating persistent_id different from another resources

**Comment by yatsukhnenko** [comment]
Current version is working correct (it doesn't crashes and valgrind shows no errors  :smile:) but there are still two issues:
1. there is a [memory leak](https://github.com/phpredis/phpredis/commit/61165f18d064ed4f22cc1deb64dbc982753d0ba6#diff-c5867178362a2abccbf1419f7363735cR1897) in PHP5
2. there is no maximum number of connections (I'll fix this in next commit)

**Comment by yatsukhnenko** [comment]
@michael-grunder here is the code which works in PHP5 and PHP7
```diff --git a/library.c b/library.c
index d799d87..73f22da 100644
--- a/library.c
+++ b/library.c
@@ -1894,15 +1894,11 @@ redis_sock_disconnect(RedisSock *redis_sock, int force TSRMLS_DC)
                 zend_string *persistent_id = strpprintf(0, "phpredis_%s:%d", ZSTR_VAL(redis_sock->host), redis_sock->port);
                 zend_resource *le = zend_hash_find_ptr(&EG(persistent_list), persistent_id);
                 if (!le) {
-#if (PHP_MAJOR_VERSION < 7)
-                    le = ecalloc(1, sizeof(*le));
-#else
-                    zend_resource res;
-                    le = &res;
-#endif
+                    zend_llist *l = pecalloc(1, sizeof(*l) + sizeof(*le), 1);
+                    zend_llist_init(l, sizeof(void *), NULL, 1);
+                    le = (void *)l + sizeof(*l);
                     le->type = le_redis_pconnect;
-                    le->ptr = pecalloc(1, sizeof(zend_llist), 1);
-                    zend_llist_init(le->ptr, sizeof(void *), NULL, 1);
+                    le->ptr = l;
                     zend_hash_str_update_mem(&EG(persistent_list), ZSTR_VAL(persistent_id), ZSTR_LEN(persistent_id), le, sizeof(*le));
                 }
                 zend_llist_prepend_element(le->ptr, &redis_sock->stream);
```
Don't you mind about hacks like that?

**Comment by axot** [comment]
Hello, we are testing `redis.pconnect.pooling_enabled` with php-fpm,

```
bash-4.4# curl localhost:10080/phpfpm_status
pool:                 www
process manager:      dynamic
start time:           10/Apr/2019:19:02:42 +0900
start since:          145
accepted conn:        8934
listen queue:         0
max listen queue:     0
listen queue len:     0
idle processes:       35
active processes:     12
total processes:      47
max active processes: 27
max children reached: 0
slow requests:        0
bash-4.4# netstat -tan | grep 6379 | wc -l
507
bash-4.4# netstat -tan | grep 6379 | grep ESTABLISHED  | wc -l
513
```

The result shows almost connection is in ESTABLISHED state, which means the pool is working?
But I do not understand why ~500 connections are in ESTABLISHED state, isn't it too many?

**Comment by axot** [comment]
Hi, should I opened an issue for this.

**Comment by michael-grunder** [comment]
Hi @axot,

Feel free to open an issue.  I haven't used connection pooling in a production environment yet, but I think @yatsukhnenko has.

Maybe it would be nice to dump some statistics about the connection pool (e.g. total cached sockets, total 'in use', etc).

That shouldn't be too tough since we can use module globals.

**Comment by axot** [comment]
Hello @michael-grunder, could you share about how to dump the statistics of connection pool? Thank you.

**Pull request merged**